### PR TITLE
feat: SolidSyslogBlockDevice abstraction with file-backed driver (S18.02)

### LINKED ISSUE #235: S18.02: Define SolidSyslogBlockDevice interface and file-backed driver
**Parent epic:** #112 (E18: Flash Storage Support)

## Scope

Introduce the `SolidSyslogBlockDevice` abstraction so `BlockSequence` (internal to FileStore) sits on top of a swappable block I/O driver instead of `SolidSyslogFile` directly. Behaviour preserved at the message/store level — integrator-observable behaviour is unchanged once wiring is updated.

**New public surface:**

- `Core/Interface/SolidSyslogBlockDevice.h` — `_Acquire` / `_Dispose` / `_Exists` / `_Read` / `_Append` / `_WriteAt` / `_Size` consumer API. See #112 for the illustrative shape.
- `Core/Interface/SolidSyslogBlockDeviceDefinition.h` — vtable struct (extension point for integrator-supplied drivers, e.g. flash).
- `Core/Interface/SolidSyslogFileBlockDevice.h` — file-backed driver. Wraps `SolidSyslogFile*` instances with sequence-numbered filenames. Acquire = "create empty file"; Dispose = "delete file" — preserves current behaviour.
- Storage injection per `CLAUDE.md` "Callback Conventions": `SolidSyslogFileBlockDeviceStorage` opaque blob.

**Wiring change:** `SolidSyslogFileStoreConfig` replaces `readFile` / `writeFile` / `pathPrefix` with a single `struct SolidSyslogBlockDevice* blockDevice` field. The integrator now constructs `SolidSyslogFileBlockDevice` separately and passes it in.

### BDD scenarios

- [ ] **No regression** — all pre-existing FileStore BDD scenarios remain green after step definitions are updated to wire `SolidSyslogFileBlockDevice` between the file handles and the FileStore.
- [ ] **Dispose only on capacity pressure** — Given a workload that does not hit capacity, When the system runs, Then no `Dispose` occurs (block files are not deleted until rotation forces it). Confirms behaviour is preserved from the pre-S18.02 baseline.

### Docs and header surface

- [ ] New headers added: `SolidSyslogBlockDevice.h`, `SolidSyslogBlockDeviceDefinition.h`, `SolidSyslogFileBlockDevice.h`.
- [ ] `SolidSyslogFileStoreConfig` updated — `readFile` / `writeFile` / `pathPrefix` removed; `blockDevice` added.
- [ ] `CLAUDE.md` "Public header audiences" — three new rows for the BlockDevice headers; `SolidSyslogFileStore.h` row updated for the config change.
- [ ] Existing examples and BDD step definitions updated for the new wiring.

## Out of scope

- Dispose-on-empty (S18.03 — purely behavioural).
- Example flash block device (S18.04).
- Real flash hardware drivers — supplied by the platform.

## References

- Parent epic E18 #112 — layered decomposition and BlockDevice illustrative shape.
- Predecessor story S18.01 — provides the layered decomposition this story builds on.
- `CLAUDE.md` "Callback Conventions" — storage-injection pattern.

### PR DESCRIPTION
**Pull request by DavidCozens**
## Summary

- Introduce `SolidSyslogBlockDevice` — block-indexed I/O with `Acquire`/`Dispose` lifecycle (Acquire/Dispose/Exists/Read/Append/WriteAt/Size). New public headers: `SolidSyslogBlockDevice.h`, `SolidSyslogBlockDeviceDefinition.h`.
- Implement `SolidSyslogFileBlockDevice` — file-backed driver wrapping two `SolidSyslogFile*` instances with sequence-numbered filenames `<prefix><NN>.log`. Storage-injected per CLAUDE.md callback-conventions.
- Repoint internal `BlockSequence` and `RecordStore` onto BlockDevice. `SolidSyslogFileStoreConfig` swaps `readFile`/`writeFile`/`pathPrefix` for one `blockDevice` field — integrators now construct the device separately and pass it in.
- File-handle ownership boundary moves outside `FileStore`: `SolidSyslogFileStore_Destroy` no longer closes files; the integrator destroys the block device explicitly. The behavioural cost is small and the contract is cleaner.
- Lazy reopen in `EnsureHandleOpenOnBlock` heals transient open failures on next op (the old fast-fail-on-Open semantics are gone — this is strictly better behaviour).
- New BDD baseline: "Block files are not disposed when capacity is not exhausted" (`Bdd/features/block_lifecycle.feature`). S18.03 will overturn this; pinned now per the issue spec.

## Slicing notes

Per the original plan, S3 (internal repoint) and S4 (public flip) were intended as separate slices. Combined into one commit because the dual-handle test setup change forced by S3 is the same change S4 needs — separating them would have meant a transitional FileBlockDevice-built-inside-FileStore hack and a temporary storage bloat. See DEVLOG.

## Test plan

- [x] gcc debug — 994 tests, 990 ran, 0 failures
- [x] clang debug — 994 tests, 990 ran, 0 failures
- [x] sanitize (ASan + UBSan) — 994 tests, 0 failures
- [x] coverage — 100% lines, 100% functions
- [x] tidy — clean
- [x] cppcheck — clean
- [x] clang-format — clean
- [x] BDD `block_lifecycle.feature` — passes (the existing capacity / tcp\_reconnect features error pre-existing on `main` due to a syslog-ng config-file permission issue in this dev environment, unrelated to this PR)
- Windows (MSVC), full BDD (Linux + Windows OTel), and OpenSSL integration are CI's responsibility per CLAUDE.md.

Closes #235.

## Summary by CodeRabbit

* **New Features**
  * Block-based storage system for store-and-forward capability with configurable block size, count limits, and discard policies (oldest/newest/halt)
  * Command-line options for block storage configuration (--max-blocks, --max-block-size)

* **Documentation**
  * Updated architecture documentation and IEC 62443 compliance references to reflect block store implementation

### CODE REVIEW

File: Core/Source/SolidSyslogFileBlockDevice.c [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+    return ready;
+}
+
+static inline const char* FormatBlockFilename(const struct SolidSyslogFileBlockDevice* device, SolidSyslogFormatterStorage* storage, size_t blockIndex)
+{
+    struct SolidSyslogFormatter* formatter = SolidSyslogFormatter_Create(storage, MAX_PATH_SIZE);
+
+    SolidSyslogFormatter_BoundedString(formatter, device->pathPrefix, MAX_PREFIX_LENGTH);
+    SolidSyslogFormatter_TwoDigit(formatter, (uint8_t) blockIndex);
+    SolidSyslogFormatter_BoundedString(formatter, FILE_EXTENSION, sizeof(FILE_EXTENSION) - 1);
```
[the remark this answers is not in the document]
Comment (DavidCozens) [reply Core/Source/SolidSyslogFileBlockDevice.c:191]: Not adopting this fix — the file-backed driver constrains indices to the two-digit on-disk sequence, but the underlying narrowing was real. Fixed in 4e8842c by adding an `IsValidBlockIndex` guard at every entry point and a companion boundary test in `SolidSyslogFileBlockDeviceTest.cpp`.

File: Tests/FileFake.c [hunk 2]
```diff
@@ -135,19 +135,19 @@ void FileFake_Destroy(void)
...
-    lastCreated->failNextWrite = true;
+    AsFake(file)->failNextWrite = true;
 }
 
-void FileFake_FailNextRead(void)
+void FileFake_FailNextRead(struct SolidSyslogFile* file)
 {
-    lastCreated->failNextRead = true;
+    AsFake(file)->failNextRead = true;
 }
```
[the remark this answers is not in the document]
Comment (DavidCozens) [reply Tests/FileFake.c:166]: Adopted in 8b31996. Replaced the unconditional `AsFake(file)->...` with a `TestAssert_Fail` early return for null. Single-fake call sites in `FileFakeTest.cpp` already pass `api`, so no caller breakage.

**Comment by DavidCozens** [comment]
Responses to the four outside-diff findings in the CodeRabbit review (couldn't be replied to inline):

**`Core/Interface/SolidSyslogFileStore.h` 33-45 — blockDevice ownership.** Adopted in 8b31996. The `blockDevice` field now carries a comment stating it is caller-owned, must outlive the FileStore, and is not destroyed by `SolidSyslogFileStore_Destroy`.

**`Core/Source/RecordStore.c` 271-299 — sent-flag read failures.** Not adopting the suggested change — the current behaviour is deliberate. Treating an unreadable sent-flag as already-sent keeps the scan moving past one bad byte; the alternative (jump cursor to fileSize, mark file corrupt) would skip every record after the first read failure and force an unnecessary rotation. The single skipped record surfaces downstream as an RFC 5424 §6.3.1 sequenceId gap, which the receiver detects. A future integrator-supplied error reporter will cover persistent media errors directly. Intent is now documented in code (8b31996) so this stays clear to future maintainers.

**`Core/Source/BlockSequence.c` 88-113 — wrapped sequence on resume.** Adopted (different shape than the suggested algorithm) in 4e8842c. `ScanForExistingBlocks` now treats the sequence space as circular, finds the absent-block region, and reads the run as its complement. New focused unit tests in `Tests/BlockSequenceTest.cpp` cover cold start, linear, wrapped, and single-block-at-boundary cases.

**`Core/Source/BlockSequence.c` 166-195 — Acquire/Dispose return ignored on rotation/discard.** Deferred to S18.04 (#237) and captured in that issue's "Design considerations (preliminary)" section. The pre-existing pattern (also present in the pre-S18.02 code, which ignored `SolidSyslogFile_Open`/`_Delete` returns) is benign for the file-backed driver because POSIX `O_CREAT` heals the missing-file case lazily on the next Append. For flash, where Acquire = erase, the right semantics need a real flash driver to inform them — co-designing with S18.04 is the right time to settle propagate-vs-retry-vs-error-report.

The "boundary test for unsupported block indices" nitpick is also addressed in 4e8842c.

File: Tests/BlockSequenceTest.cpp [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+    LONGS_EQUAL(1, BlockSequence_WriteSequence(&sequence));
+}
+
+TEST(BlockSequenceScan, ResumesWrappedSingleBlockAtBoundary)
+{
+    existing = {99, 0};
+    BlockSequence_Open(&sequence);
+    LONGS_EQUAL(99, BlockSequence_ReadSequence(&sequence));
+    LONGS_EQUAL(0, BlockSequence_WriteSequence(&sequence));
+}
```
[the remark this answers is not in the document]
Comment (DavidCozens) [reply Tests/BlockSequenceTest.cpp:156]: Addressed in 556b896 — `CHECK_TRUE(BlockSequence_Open(&sequence))` added at the start of all five resume tests. Good catch; the zero-init fixture would otherwise have masked an Open/Acquire regression.

File: Core/Source/SolidSyslogBlockStore.c [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+        RecordStore_FindFirstUnsent(&blockStore->recordStore, BlockSequence_BlockDevice(&blockStore->blockSequence),
+                                    BlockSequence_ReadSequence(&blockStore->blockSequence), BlockSequence_WritePosition(&blockStore->blockSequence), &corrupt);
+
+    BlockSequence_SetReadCursor(&blockStore->blockSequence, cursor);
+
+    if (corrupt)
+    {
+        BlockSequence_MarkWriteBlockCorrupt(&blockStore->blockSequence);
+    }
+}
```
[the remark this answers is not in the document]
Comment (DavidCozens) [reply Core/Source/SolidSyslogBlockStore.c:132]: Fixed in 8b13d2f — `ResumeFromExistingBlock` now bounds the scan by `SolidSyslogBlockDevice_Size(device, readSequence)` instead of `WritePosition`. You're right that on multi-block resume the read block is a closed earlier block whose size is independent of the write block's fill level. Confirmed the regression test (`ResumeFindsUnsentInClosedReadBlockWhenWriteBlockPartial`) fails without the fix and passes with it. Also removed the now-unused `BlockSequence_WritePosition` accessor (was its only caller).

**Comment by DavidCozens** [comment]
Sweep through the outstanding CodeRabbit comments — pushed two follow-up commits and deferred one item to S18.03.

**Addressed in 556b896 (test hygiene):**
- `Tests/Example/ExampleCommandLineTest.cpp` — renamed `DefaultMaxFiles` → `DefaultMaxBlocks` to match the renamed `options.maxBlocks` field.
- `Tests/SolidSyslogBlockStoreTest.cpp:505-533` — renamed `MaxFiles*` / `MaxFileSize*` test names → `MaxBlocks*` / `MaxBlockSize*`.
- `Tests/SolidSyslogFileBlockDeviceTest.cpp` — added `ReacquireTruncatesExistingBlock` covering the truncate-on-reacquire contract that `BlockSequence_RotateToNextBlock` relies on.
- `Tests/BlockSequenceTest.cpp` — `CHECK_TRUE(BlockSequence_Open(&sequence))` added to all five resume tests.

**Fixed in 8b13d2f (production bug + regression test):**
- `Core/Source/SolidSyslogBlockStore.c` — `ResumeFromExistingBlock` now bounds the scan by `SolidSyslogBlockDevice_Size(device, readSequence)` rather than `WritePosition`. New `ResumeFindsUnsentInClosedReadBlockWhenWriteBlockPartial` test fails without the fix and passes with it. The unused `BlockSequence_WritePosition` accessor was removed.

**Deferred to S18.03 #236:**
- `Core/Source/BlockSequence.c` — `RotateToNextBlock` and `DiscardOldestBlock` not propagating `Acquire` / `Dispose` failure: real concern, but it changes the shape of `PrepareForWrite`'s contract and overlaps S18.03's empty-check-on-Acquire work. #236 has been updated with the explicit success-only-state-advancement scope item plus matching BDD acceptance criteria.

**Already addressed in earlier commits — leaving as historical context:**
- `Core/Source/BlockSequence.c:88-113` (wrap-around in `ScanForExistingBlocks`) — handled by `LocateRunBoundaries` in 4e8842c; the doc comment at lines 107-112 describes the gap-detection approach, and `BlockSequenceTest.cpp` has `ResumesWrappedSequenceRangeCorrectly` / `ResumesWrappedSingleBlockAtBoundary` covering it.
- `Core/Interface/SolidSyslogFileStore.h:33-45` (blockDevice ownership/lifetime doc) — the rename to `SolidSyslogBlockStore.h` carried the doc forward; current header lines 33-37 say *Caller-owned: must outlive the BlockStore. SolidSyslogBlockStore_Destroy does NOT destroy the block device — that is the integrator's responsibility.*

**Declining — see in-tree design rationale:**
- `Core/Source/RecordStore.c:271-299` (treat sent-flag read failures as corruption instead of "sent") — the comment at lines 294-301 explicitly defends the current behaviour and rejects the proposed alternative: *"a sent-flag we cannot read is treated as already-sent so the scan keeps walking the block. The alternative — stop scanning, refuse to advance — would jam the logger on one bad byte and skip every record that follows in the same block."* Single skipped records surface to the SIEM via the `MetaSd` sequenceId gap (RFC 5424 §7.3.1). I'd rather keep that design and surface persistent media errors via the integrator-supplied error reporter when that path lands.
